src/engine_plugin.cpp

```diff
@@ -135,6 +135,28 @@ class engine_plugin_impl : std::enable_shared_from_this<engine_plugin_impl> {
          return silkworm::db::read_canonical_header(txn, head_num);
       }
 
+      std::optional<silkworm::BlockHeader> find_block_with_valid_eos_id(uint64_t start_height) {
+         // Search for the last block with a valid eos id in it.
+         // The search is not that optimized. However, this function will only be called during initialization process
+         // in rara cases and the gap in eos will not be too large in most cases. Therefore, it should be fine to leave 
+         // the search process simple.
+         SILK_INFO << "Search for block containing a valid eos id. Start from:" << "#" << start_height;
+         silkworm::db::ROTxn txn(db_env);
+         do {
+            auto res = silkworm::db::read_canonical_header(txn, start_height);
+            if(!res) return {};
+
+            for (int i = 0; i < 32; ++i) {
+               if (res->prev_randao.bytes[i]) {
+                  SILK_INFO << "Found block at: " << "#" << start_height;
+                  return res;
+               }
+            }
+         }
+         while(--start_height > 0);
+         return {};
+      }
+
```

Comment:
// Alternative implementation
```c++
std::optional<silkworm::BlockHeader> res;
while(start_height>0) {
  auto res = silkworm::db::read_canonical_header(txn, start_height);
  if(res && !is_zero(res->prev_randao)) {
     break;
  }
 --start_height;
}
return res;

Reply:
fixed as this

Reply:
Refactor to do while again to handle the genesis case properly.